### PR TITLE
Grant read right to cloud storage bucket, create source representation and replica instance.

### PR DESCRIPTION
**Pull request by cdrake**
This request adds granting read rights to an email (necessary for replication initialization), creating the source representation of the external master and replica instance.

Updated tests for update cloud storage methods.

### CODE REVIEW

File: CloudSQLReplicator/google/cloud/pontem/sql/replicator/util/gcp_api_util.py [hunk 1]
```diff
@@ -24,7 +24,16 @@
 
 from google.cloud.pontem.sql import replicator
 
-DEFAULT_TIER = 'db-n1-standard-2'
+
+STORAGE_SCOPE = ['https://www.googleapis.com/auth/devstorage.read_write']
```
Comment (EricBeach) [inline CloudSQLReplicator/google/cloud/pontem/sql/replicator/util/gcp_api_util.py]: Is this a constant anywhere that you can depend on? If not, no worries. If it is, when we should use the constant.

Comment (EricBeach) [inline CloudSQLReplicator/google/cloud/pontem/sql/replicator/util/gcp_api_util.py]: STORAGE_SCOPE is a tad vague; CLOUDSQL_STORAGE or GCS_STORAGE_SCOPE etc?

File: CloudSQLReplicator/google/cloud/pontem/sql/replicator/util/gcp_api_util.py [hunk 2]
```diff
@@ -96,6 +103,116 @@ def create_cloudsql_instance(database_instance_body=None,
...
+      db_version=DEFAULT_2ND_GEN_DB_VERSION,
+      region=DEFAULT_2ND_GEN_REGION,
+      source_representation_name=None,
+      source_representation_body=None,
+      project=None,
+      credentials=None):
+  """Creates a source representation of an external master.
+
+
+  If source representation is included source_representation_name, ip_address,
```
Comment (EricBeach) [inline CloudSQLReplicator/google/cloud/pontem/sql/replicator/util/gcp_api_util.py]: do you need a comma after "included"?

File: CloudSQLReplicator/google/cloud/pontem/sql/replicator/util/gcp_api_util.py [hunk 3]
```diff
@@ -96,6 +103,116 @@ def create_cloudsql_instance(database_instance_body=None,
...
+  Args:
+    source_representation_name (str): The instance name of the external master.
+    ip_address (str): The ip address of the external master.
+    port (str): Port that will be used for replication.
+    region (str): Region source representation will be created.
+    database_version (str): MySQL database version
+    source_representation_body (JSON): Creation options for source
+      representation.
+    project (str): Project ID where replica will be created.
+      credentials (google.auth.Credentials): Credentials to authorize client.
```
Comment (EricBeach) [inline CloudSQLReplicator/google/cloud/pontem/sql/replicator/util/gcp_api_util.py]: should the URL be indented?

File: CloudSQLReplicator/google/cloud/pontem/sql/replicator/util/gcp_api_util.py [hunk 4]
```diff
@@ -96,6 +103,116 @@ def create_cloudsql_instance(database_instance_body=None,
...
+  Returns:
+      JSON: response from sqladmin.instances().insert() call.
+  """
+  default_source_representation_body = {
+    'name': source_representation_name or 'external-mysql-representation-{}'.format(uuid.uuid4()),
+    'databaseVersion': db_version,
+    'region': region,
+    'onPremisesConfiguration': {
+      "kind": "sql#onPremisesConfiguration",
+      "hostPort": '{}:{}'.format(ip_address, port)
```
Comment (EricBeach) [inline CloudSQLReplicator/google/cloud/pontem/sql/replicator/util/gcp_api_util.py]: why alternating between single and double quotes?

File: CloudSQLReplicator/google/cloud/pontem/sql/replicator/util/gcp_api_util.py [hunk 5]
```diff
@@ -205,12 +321,32 @@ def delete_blob(bucket_name,
...
+                                project=None,
+                                credentials=None
+                           ):
+  """Grants read access to a specific bucket.
+
+    Args:
+        bucket_name (str): name of bucket that the user will be able to read.
+        email (str): email of user who will be granted read privileges.
+        project (str): Project ID where blob will be deleted.
+       credentials (google.auth.Credentials): credentials to authorize client.
```
Comment (EricBeach) [inline CloudSQLReplicator/google/cloud/pontem/sql/replicator/util/gcp_api_util.py]: spacing for credentials -- should be moved right one space. shouldn't the linter catch this? is it not being run?

File: CloudSQLReplicator/tests/util/gcp_api_util_test.py [hunk 6]
```diff
@@ -17,24 +17,31 @@
...
 from mock import MagicMock
-from google.auth import compute_engine
+
+import google.auth
 from google.cloud import storage
+from google.oauth2 import credentials
 
 from google.cloud.pontem.sql.replicator.util import gcp_api_util
 
+STORAGE_SCOPE = ['https://www.googleapis.com/auth/devstorage.read_write']
```
Comment (EricBeach) [inline CloudSQLReplicator/tests/util/gcp_api_util_test.py]: was this already defined once in another file? should we be sharing?

File: CloudSQLReplicator/google/cloud/pontem/sql/replicator/util/gcp_api_util.py [hunk 7]
```diff
@@ -96,6 +103,116 @@ def create_cloudsql_instance(database_instance_body=None,
...
+  """Provisions a Cloud SQL Replica instance.
+
+    Will create a second generation replica by default, specify tier and
+    region if creating a first generation replica.
+
+    If replica_instance_body is supplied, master_instance_name, dumpfile_path
+      replica_user, replica_pwd, and replica_instance_name will be ignored.
+
+    Args:
+      master_instance_name (str): Instanc name of master  that will be
```
Comment (ldanielmadariaga) [inline CloudSQLReplicator/google/cloud/pontem/sql/replicator/util/gcp_api_util.py]: Nits: typo in instance double space between master and that

File: CloudSQLReplicator/google/cloud/pontem/sql/replicator/util/gcp_api_util.py [hunk 8]
```diff
@@ -96,6 +103,116 @@ def create_cloudsql_instance(database_instance_body=None,
...
+      replica_instance_name (str): Name of replica instance to create.
+      replica_instance_body (JSON): Options for replica instance creation.
+      project (str): Project ID where replica will be created.
+      credentials (google.auth.Credentials): Credentials to authorize client
+
+    Returns:
+      JSON: response from sqladmin.instances().insert() call
+  """
+
+  default_replica_intance_body = {
```
Comment (ldanielmadariaga) [inline CloudSQLReplicator/google/cloud/pontem/sql/replicator/util/gcp_api_util.py]: typo: instance

File: CloudSQLReplicator/google/cloud/pontem/sql/replicator/util/gcp_api_util.py (hunk 1 above)
Comment (cdrake) [reply CloudSQLReplicator/google/cloud/pontem/sql/replicator/util/gcp_api_util.py]: I have just seen the literal in code examples.

Comment (cdrake) [reply CloudSQLReplicator/google/cloud/pontem/sql/replicator/util/gcp_api_util.py]: Changed to GCP_STORAGE_SCOPE.

File: CloudSQLReplicator/google/cloud/pontem/sql/replicator/util/gcp_api_util.py [hunk 9]
```diff
@@ -96,6 +103,116 @@ def create_cloudsql_instance(database_instance_body=None,
...
+
+  return response
+
+
+def create_cloudsql_replica_instance(
+      master_instance_name,
+      dumpfile_path,
+      replica_user,
+      replica_pwd,
+      replica_instance_name=None,
```
Comment (nagarkar) [inline CloudSQLReplicator/google/cloud/pontem/sql/replicator/util/gcp_api_util.py]: Do you want to have tier, region, etc as parameters so you can override the constants you are passing in by default?

File: CloudSQLReplicator/google/cloud/pontem/sql/replicator/util/gcp_api_util.py (hunk 3 above)
Comment (cdrake) [reply CloudSQLReplicator/google/cloud/pontem/sql/replicator/util/gcp_api_util.py]: Did not mean to leave the URL in the docstring.

File: CloudSQLReplicator/google/cloud/pontem/sql/replicator/util/gcp_api_util.py (hunk 4 above)
Comment (cdrake) [reply CloudSQLReplicator/google/cloud/pontem/sql/replicator/util/gcp_api_util.py]: Fixed.

File: CloudSQLReplicator/google/cloud/pontem/sql/replicator/util/gcp_api_util.py [hunk 10]
```diff
@@ -81,10 +89,9 @@ def create_cloudsql_instance(database_instance_body=None,
 
   default_credentials, default_project = google.auth.default()
   default_database_intance_body = {
-    'project': default_project,
     'name': 'cloudsql-db-{}'.format(uuid.uuid4()),
     'settings': {
-      'tier': DEFAULT_TIER
+      'tier': DEFAULT_2ND_GEN_TIER
```
Comment (nagarkar) [inline CloudSQLReplicator/google/cloud/pontem/sql/replicator/util/gcp_api_util.py]: This is not strictly related to this code review, but are you handling retries in this function when you call service.instances().insert(..)?

Comment (cdrake) [reply CloudSQLReplicator/google/cloud/pontem/sql/replicator/util/gcp_api_util.py]: No, but I will add retry later.

File: CloudSQLReplicator/google/cloud/pontem/sql/replicator/util/gcp_api_util.py (hunk 9 above)
Comment (cdrake) [reply CloudSQLReplicator/google/cloud/pontem/sql/replicator/util/gcp_api_util.py]: Addressed in our conversation (i.e. use replica_instance_body to override defaults).

File: CloudSQLReplicator/google/cloud/pontem/sql/replicator/util/gcp_api_util.py (hunk 8 above)
Comment (cdrake) [reply CloudSQLReplicator/google/cloud/pontem/sql/replicator/util/gcp_api_util.py]: Fixed.
